hibernate-core/src/main/java/org/hibernate/engine/internal/TwoPhaseLoad.java

```diff
@@ -206,62 +206,91 @@ public static void initializeEntityEntryLoadedState(
 		String entityName = persister.getEntityName();
 		String[] propertyNames = persister.getPropertyNames();
 		final Type[] types = persister.getPropertyTypes();
-		
-		final GraphImplementor<?> fetchGraphContext = session.getFetchGraphLoadContext();
-		
-		for ( int i = 0; i < hydratedState.length; i++ ) {
-			final Object value = hydratedState[i];
-			if ( debugEnabled ) {
-				LOG.debugf(
-					"Processing attribute `%s` : value = %s",
-					propertyNames[i],
-					value == LazyPropertyInitializer.UNFETCHED_PROPERTY ? "<un-fetched>" : value == PropertyAccessStrategyBackRefImpl.UNKNOWN ? "<unknown>" : value
-				);
-			}
 
-			if ( value == LazyPropertyInitializer.UNFETCHED_PROPERTY ) {
+		GraphImplementor fetchGraphContext = session.getFetchGraphLoadContext();
+		if ( fetchGraphContext != null && !fetchGraphContext.appliesTo( entity.getClass() ) ) {
+			LOG.warnf( "Entity graph specified is not applicable to the entity [%s]. Ignored.", entity);
+			fetchGraphContext = null;
+			session.setFetchGraphLoadContext( null );
```

Comment:
just a minor but I think the `session.setFetchGraphLoadContext( null );` is redundant

Reply:
I think it is necessary because without setting it, the downstream will still try to locate some property as per the context grabbed from `session.getFetchGraphLoadContext()`, not from the local variable of `fetchGraphContext`. See the `org.hibernate.engine.internal.TwoPhaseLoad#isEagerFetchGraph(SharedSessionContractImplementor session, String associationName, Type associationType)` for details


I am sorry the current fetch entity graph implementation is so difficult to understand (and maintain). Luckily its status is improved much more in v6.